### PR TITLE
publish the sub test result for guest upgrade tests in openQA

### PR DESCRIPTION
**Pull request by Julie-CAO**
The guest upgrade test may has 1-2 sub testcases, fg.fv and pv for xen. Currently when one sub test fails and another succeeds,  the test just fails without telling which ones are failing in guest_upgrade_run. the fix will list the result for each sub tests whatever they are failing or successful.

Overall guest upgrade result is:
Guest Name --- Result --- Reason
sles-12-sp3-64-fv-def-net --- Fail --- Guest upgrade without reboot fail.
sles-12-sp3-64-pv-def-net --- Pass

- The Trello card: https://trello.com/c/tOdrk5oq/212-p3-openqaguestupgradelog-parser-couldnt-present-the-sub-test-result
- Needles: no needles change
- Verification run: http://10.67.18.247/tests/8#

@alice-suse @XGWang0 @xguo @waynechen55

### CODE REVIEW

**Comment by Julie-CAO** [comment]
@alice-suse @XGWang0 @xguo @waynechen55

File: tests/virt_autotest/guest_upgrade_run.pm [hunk 1]
```diff
@@ -34,6 +34,28 @@ sub get_script_run {
...
+
+    foreach (split("\n", $rough_result)) {
+        if ($_ =~ /(\S+)\s+\-{3}\s+(PASS|FAIL|SKIP|TIMEOUT)(\s+\-{3}\s+(.*))*/ig) {
+            my ($testcase_name, $status, $error) = ($1, $2, $4);
+            if ($status =~ /pass/i) {
+                $status = "PASSED";
+            }
+            elsif ($status =~ /fail/i) {
+                $status = "FAILED";
+            }
```
Comment (okurz) [inline tests/virt_autotest/guest_upgrade_run.pm]: How about just `$result->{$testcase_name}->{status} = $status =~ /pass/i ? 'PASSED' : 'FAILED';` ?

File: tests/virt_autotest/guest_upgrade_run.pm [hunk 2]
```diff
@@ -34,6 +34,28 @@ sub get_script_run {
...
 
+sub analyzeResult {
+    my ($self, $text) = @_;
+    my $result;
+    $text =~ /Overall guest upgrade result is:(.*)Test done/s;
+    my $rough_result = $1;
+
+    foreach (split("\n", $rough_result)) {
+        if ($_ =~ /(\S+)\s+\-{3}\s+(PASS|FAIL|SKIP|TIMEOUT)(\s+\-{3}\s+(.*))*/ig) {
+            my ($testcase_name, $status, $error) = ($1, $2, $4);
```
Comment (rwx788) [inline tests/virt_autotest/guest_upgrade_run.pm]: Just a small one, we could use named regexp groups. 
`/(?<testcase_name>\S+)\s+\-{3}\s+(PASS|FAIL|SKIP|TIMEOUT)(\s+\-{3}\s+(.*))*` and then you can access it with  `$+{testcase_name}`. Be aware that this hash will get overwritten after you call other match, so you still need variables assignment.

**Comment by Julie-CAO** [comment]
edit codes locally per suggestions here. will submit after local testing.

**Comment by Julie-CAO** [comment]
If everything is ok, I will merge the commits into one.

File: tests/virt_autotest/guest_upgrade_run.pm [hunk 3]
```diff
@@ -34,6 +34,24 @@ sub get_script_run {
...
+sub analyzeResult {
+    my ($self, $text) = @_;
+    my $result;
+    if ($text !~ /Overall guest upgrade result is:(.*)Test done/s) {
+        die "The sub tests results are not listed in the test output. \n";
+    }
+    my $rough_result = $1;
+
+    foreach (split("\n", $rough_result)) {
+        if ($_ =~ /(?<testcase_name>\S+)\s+\-{3}\s+(PASS|FAIL|SKIP|TIMEOUT)(\s+\-{3}\s+(.*))?/ig) {
```
Comment (rwx788) [inline tests/virt_autotest/guest_upgrade_run.pm]: Changes LGTM. The only thing I'm thinking about, is that we have multiple overrides of 'analyzeResult' and they all are 90% same, and have different statuses e.g. `PASS|FAIL|SKIP|TIMEOUT` vs `PASSED|FAILED|SKIPPED|TIMEOUT` vs `PASSED|FAILED|SKIPPED`. Is this done on purpose? I'm not saying that we have to address this in this PR, but seems that it should be possible to unify this code and have it in single place. WDYT?

Comment (Julie-CAO) [reply tests/virt_autotest/guest_upgrade_run.pm]: Yes, we have multiple 'analyzeResult' in our tests as you said. the test outputs are some kind of different among tests. We should consider to re-organize and standardize the log output for our the tests in our group. We will think about it in our team. thanks.

**Comment by Julie-CAO** [comment]
The 5 commits have been merged into one. My commit succeeded in build with Perl5.26 while failed in the one with Perl5.18 in your CI checks. what should I do?

**Comment by okurz** [comment]
Looks like the test on perl5.18 failed on installing dependencies, that is weird. I retriggered as https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/448775610 . Let's see if this can help. As our production infrastructure runs perl5.18 we should ensure this works there.

**Comment by rwx788** [comment]
@Julie-CAO bold answer is to make it work for both =). I've taken a look and it's quite strange, as it's about not installed XML::LibXML, which is already in cpan file. @okurz could you please take a quick look?

**Comment by okurz** [comment]
tests passed now, we can merge :)

**Pull request merged**